File: triangle.rb
class Triangle
  def initialize(sides)
    @sides = sides
  end

  def equilateral?
    num_of_uniq_sides == 1 && valid_triangle?
  end

  def isosceles?
    num_of_uniq_sides <= 2 && valid_triangle?
  end

  def scalene?
    num_of_uniq_sides == 3 && valid_triangle?
  end

  private

  attr_reader :sides

  def num_of_uniq_sides
    sides.uniq.size
  end

  def valid_triangle?
      sides.size == 3 && sides.min.positive? && sides.permutation.all? { |a, b, c| a < b + c }
  end
end

puts Triangle.new([0, 0, 0]).equilateral?
puts Triangle.new([4, 4, 4]).equilateral?

File: python/triangle.py
def valid_triangle(sides):
    if not any(sides): return False
    if not len(sides) == 3: return False
    a, b, c = sides
    return not (a + b < c or b + c < a or c + a < b)

def equilateral(sides):
    if not valid_triangle(sides): return False
    return len(set(sides)) == 1

def isosceles(sides):
    if not valid_triangle(sides): return False
    a, b, c = sides
    if a == b or a == c or b == c: return True
    return False

def scalene(sides):
    if not valid_triangle(sides): return False
    return len(set(sides)) == 3

print(equilateral([2, 2, 2]))